[FIX] sale_coupon_multi_use: sale confirm

oerp-odoo opened the pull request:
[FIX] sale_coupon_multi_use: sale confirm

Coupons are marked as consumed twice. First time when coupon is added
and second when sale order is confirmed.

For that, multi-use coupon must not be marked as consumed when
confirming order (if it is still valid by multi-use rules).

Also had to redesign the way multi use coupons are checked on write,
because during confirmation it can update multiple coupons and old
implementation was able to handle one coupon at a time.

On top of that, simplified multi-use coupon handling. Now correct amount
is intercepted when discount_fixed_amount is computed for reward line.

This gives multiple advantages:
    - no need to adjust sale order line discount afterwards.
    - _compute_program_amount handles currency conversion, so multi-use
        consumed amount now also takes care of that too.
    - simpler consumption line creation and checking if coupon is fully
        consumed (no need to compare delta with discount that would need
        to be computed from sale order).

oerp-odoo committed:
[REF] sale_coupon_multi_use: LIC/naming/misc/lang

oerp-odoo committed:
[FIX] sale_coupon_multi_use: sale confirm

Coupons are marked as consumed twice. First time when coupon is added
and second when sale order is confirmed.

For that, multi-use coupon must not be marked as consumed when
confirming order (if it is still valid by multi-use rules).

Also had to redesign the way multi use coupons are checked on write,
because during confirmation it can update multiple coupons and old
implementation was able to handle one coupon at a time.

On top of that, simplified multi-use coupon handling. Now correct amount
is intercepted when discount_fixed_amount is computed for reward line.

This gives multiple advantages:
    - no need to adjust sale order line discount afterwards.
    - _compute_program_amount handles currency conversion, so multi-use
        consumed amount now also takes care of that too.
    - simpler consumption line creation and checking if coupon is fully
        consumed (no need to compare delta with discount that would need
        to be computed from sale order).

StephaneMangin commented on sale_coupon_multi_use/i18n/fr.po:
```diff
@@ -63,12 +66,14 @@ msgstr "Montant restant"
 #: model:ir.model.fields,field_description:sale_coupon_multi_use.field_sale_coupon_consumption_line__id
 msgid "ID"
-msgstr "ID"
+msgstr ""
```
```suggestion
msgstr "Identifiant"
```

StephaneMangin commented on sale_coupon_multi_use/models/sale_coupon.py:
```diff
@@ -64,32 +55,25 @@ def _get_related_sale_order_line(self, sale_order):
+    def _prepare_consumption_line(self, sale_order_line):
         self.ensure_one()
-        line = self._get_related_sale_order_line(sale_order)
+        # line = self._get_related_sale_order_line(sale_order)
```
Maybe you could just forget this statement.

oerp-odoo replied:
Yeah, leftover after refactoring,

StephaneMangin reviewed (approved):
LG

i-vyshnevska commented on sale_coupon_multi_use/models/sale_order.py:
```diff
@@ -0,0 +1,14 @@
+
+    _inherit = "sale.order"
+
+    def action_confirm(self):
```
do we need to pass order in action_cancel and drop all consumption lines?

oerp-odoo replied:
Hm, looks like yes. Need to drop lines, because coupon is cleared

i-vyshnevska commented on sale_coupon_multi_use/models/sale_coupon.py:
```diff
@@ -28,32 +37,15 @@ def _compute_discount_fixed_amount_delta(self):
-        coupon_line.price_unit += amount_to_adjust
+    def _filter_multi_use_triggered(self, vals):
+        # Indicating for coupon to be consumed
+        if vals.get("state") == "used":
```
should we also consider coupons in "new"?

i-vyshnevska commented on sale_coupon_multi_use/models/sale_coupon.py:
```diff
@@ -63,32 +55,24 @@ def _get_related_sale_order_line(self, sale_order):
             "coupon_id": self.id,
-            "amount": amount_consumed,
-            "sale_order_line_id": line.id,
+            "amount": abs(sale_order_line.price_total),
```
do we need to check if the amount of coupon is not exceeded before line creation 
like if the amount on coupon left is 20 and order_line.price_total is 50
IMO this should be filtered earlier than on creation but I see filtering on the program not on the coupon, maybe I missing smth

oerp-odoo replied:
The way it works now, it finds current discount delta (remaining amount for discount). Then it compares with existing sale order total amount. And it uses the lower amount -> so it won't apply higher discount than order itself and it won't apply higher discount than there is discount delta (remaining amount).

And that line created is actually used to finalize consumption and return flag whether coupon is fully consumed. In previous implementation, it worked differently, where it would check what kind of amount to use for consumption line, but it was actually more problematic, so I redesigned to intercept earlier (when sale_coupon module actually tries to add discount on sale order line). This have multiple advantages. You can read commit message for that.

oerp-odoo committed:
[FIX] sale_coupon_multi_use: action_cancel

Make sure related consumption lines are removed if coupon is being
reset (made valid again), e.g. SO is being cancelled.
